tracer.go

```diff
@@ -1,48 +1,72 @@
 package zipkin
 
 import (
+	"context"
+	"sync/atomic"
 	"time"
 
-	"github.com/openzipkin/zipkin-go/kind"
+	"github.com/openzipkin/zipkin-go/idgenerator"
+	"github.com/openzipkin/zipkin-go/model"
+	"github.com/openzipkin/zipkin-go/propagation"
+	"github.com/openzipkin/zipkin-go/reporter"
 )
 
 // Tracer is our Zipkin tracer implementation.
 type Tracer struct {
-	options TracerOptions
+	noop                 int32
+	localEndpoint        *model.Endpoint
+	sharedSpans          bool
+	sampler              Sampler
+	generate             idgenerator.IDGenerator
+	defaultTags          map[string]string
+	unsampledNoop        bool
+	extractFailurePolicy ExtractFailurePolicy
+	reporter             reporter.Reporter
 }
 
 // NewTracer returns a new Zipkin Tracer.
-func NewTracer(options ...TracerOption) (*Tracer, error) {
+func NewTracer(reporter reporter.Reporter, options ...TracerOption) (*Tracer, error) {
 	// set default tracer options
-	opts := &TracerOptions{
+	t := &Tracer{
 		sharedSpans: true,
 		sampler:     alwaysSample,
-		generate:    &RandomID64{},
+		generate:    idgenerator.NewRandom64(),
 		defaultTags: make(map[string]string),
-		transport:   &NoopTransport{},
+		reporter:    reporter,
 	}
 
 	// process functional options
 	for _, option := range options {
-		if err := option(opts); err != nil {
+		if err := option(t); err != nil {
 			return nil, err
 		}
 	}
 
-	return &Tracer{options: *opts}, nil
+	return t, nil
 }
 
-// StartSpan creates and starts a span
-func (t *Tracer) StartSpan(
-	name string, kind kind.Type, options ...SpanOption,
-) Span {
+// StartSpanFromContext creates and starts a span using the span found in
+// context as parent. If no parent span is found a root span is created.
+func (t *Tracer) StartSpanFromContext(ctx context.Context, name string, options ...SpanOption) (Span, context.Context) {
+	if parentSpan := SpanFromContext(ctx); parentSpan != nil {
+		options = append(options, Parent(parentSpan.Context()))
+	}
+	span := t.StartSpan(name, options...)
+	return span, NewContext(ctx, span)
+}
+
+// StartSpan creates and starts a span.
+func (t *Tracer) StartSpan(name string, options ...SpanOption) Span {
```

Comment:
This is kinda very OT specific. I'd drop this method and allow spans to start theirselves.

Reply:
Sorry I disagree.

Reply:
I mean, brave does not do this, the span is created at one stage an started in another stage: https://github.com/openzipkin/brave/blob/master/brave/src/main/java/brave/Tracer.java#L212

Reply:
yes which aligns well with how Java code is written given the various frameworks/hooks to use; but looks awkward from a Go developers perspective. Splitting it up just adds a burden to somebody instrumenting code by now having to call multiple functions while I have not seen a strong use case for starting span timing after span creation.